lms/djangoapps/courseware/views/views.py

```diff
@@ -749,40 +742,118 @@ def _progress(request, course_key, student_id):
         'staff_access': staff_access,
         'student': student,
         'passed': is_course_passed(course, grade_summary),
-        'show_generate_cert_btn': show_generate_cert_btn,
         'credit_course_requirements': _credit_course_requirements(course_key, student),
-        'missing_required_verification': missing_required_verification,
-        'certificate_invalidated': False,
-        'enrollment_mode': enrollment_mode,
+        'certificate_data': _get_cert_data(student, course, course_key, is_active, enrollment_mode)
     }
 
-    if show_generate_cert_btn:
-        # If current certificate is invalidated by instructor
-        # then show the certificate invalidated message.
-        context.update({
-            'certificate_invalidated': certs_api.is_certificate_invalid(student, course_key)
-        })
+    with outer_atomic():
+        response = render_to_response('courseware/progress.html', context)
+
+    return response
 
-        cert_status = certs_api.certificate_downloadable_status(student, course_key)
-        context.update(cert_status)
-        # showing the certificate web view button if feature flags are enabled.
+
+def _get_cert_data(student, course, course_key, is_active, enrollment_mode):
+    """Returns students course certificate related data.
+
+    Arguments:
+        student (User): Student for whom certificate to retrieve.
+        course (Course): Course object for which certificate data to retrieve.
+        course_key (CourseKey): Course identifier for course.
+        is_active (Bool): Boolean value to check if course is active.
+        enrollment_mode (String): Course mode in which student is enrolled.
+
+    Returns:
+        returns dict if course certificate is available else None.
+    """
+    show_generate_cert_btn = (
+        is_active and CourseMode.is_eligible_for_certificate(enrollment_mode)
+        and certs_api.cert_generation_enabled(course_key)
+    )
+
+    show_cert_web_view = False
+    download_url = None
+    cert_web_view_url = None
+
+    if enrollment_mode == CourseMode.AUDIT:
+        return Cert_Data(
+            CertificateStatuses.audit_passing,
+            'Your enrollment: Audit track',
+            'You are enrolled in the audit track for this course. The audit track does not include a certificate.',
+            download_url,
+            show_cert_web_view,
+            cert_web_view_url
```

Comment:
I think this will be clearer if you pass the `download_url`, `show_cert_web_view`, and `cert_web_view_url` values directly (as keyword arguments), rather than setting them before the `if` block.

Something like: 


    if enrollment_mode == CourseMode.AUDIT:
        return Cert_Data(
            CertificateStatuses.audit_passing,
            'Your enrollment: Audit track',
            'You are enrolled in the audit track for this course. The audit track does not include a certificate.',
            download_url=None,
            show_cert_web_view=False,
            cert_web_view_url=None
    )